File: BOJ/27_동적 계획법과 최단거리 역추적/12852_1로만들기2.py
N = int(input())

dp = [0,0,1,1]
path = [0,1,1,1]

for i in range(4,N+1) :
    dp.append(dp[i-1] + 1)
    path.append(i-1)

    if i % 3 == 0 :
        if dp[i//3]+1 < dp[i] :
            dp[i] = dp[i//3]+1
            path[i] = i//3
    if i % 2 == 0 :
        if dp[i//2]+1 < dp[i] :
            dp[i] = dp[i//2]+1
            path[i] = i//2

print(dp[N])
print(N,end=' ')
while N != 1 :
    print(path[N],end=' ')
    N = path[N]

File: BOJ/27_동적 계획법과 최단거리 역추적/12852_1로만들기2.cpp
#include <cstdio>

#define MAXN 1000000
int dp[MAXN+1], path[MAXN+1];
int N;
int main() {

    scanf("%d",&N);

    dp[0] = 0; path[0] = 0;
    dp[1] = 0; path[1] = 0;
    dp[2] = 1; path[2] = 1;
    dp[3] = 1; path[3] = 1;

    for(int i=4; i<=N; i++) {
        dp[i] = dp[i-1] + 1;
        path[i] = i-1;

        if(i % 3 == 0) {
            if(dp[i/3]+1 < dp[i]) {
                dp[i] = dp[i/3]+1;
                path[i] = i / 3;
            }
        }
        if(i % 2 == 0){
            if(dp[i/2]+1 < dp[i]) {
                dp[i] = dp[i/2]+1;
                path[i] = i / 2;
            }
        }
    }

    printf("%d\n",dp[N]);
    printf("%d ",N);
    while(N != 1) {
        printf("%d ",path[N]);
        N = path[N];
    }
    
    return 0;
}